File: programmers/algorithm_data_structure/greedy/섬 연결하기.py

def my_solution(n, costs):
    '''
    매번 연결된 섬과 연결되지 않은 섬을 연결하는 다리 목록을 계산하여
    그중 최소 비용의 다리를 선택해 건설하는 Greedy 알고리즘.

    -------------------------------
    [ Prim's Algorithm ] 에 해당한다.
    https://ko.wikipedia.org/wiki/프림_알고리즘
    => 아래 best_prim_solution() 함수 방법대로 구현하자
    -------------------------------
    다른 코드에 비해 시간 효율성이 안좋다.
    (매번 목록을 새로 만들기 때문)
    '''
    answer = 0

    remain_n = set(range(1, n))  
    while remain_n:
        avail_bridge = [cost for cost in costs
                        if (cost[0] in remain_n and cost[1] not in remain_n)
                        or (cost[0] not in remain_n and cost[1] in remain_n)]

        i1, i2, c = min(avail_bridge, key=lambda x:x[2])
        remain_n -= set([i1, i2])
        answer += c

    return answer


def best_prim_solution(n, costs):
    '''
    좀 더 실행 시간이 짧은 코드
    (타인이 작성한 코드를 정리함)
    `prim algorithm`도 정렬을 해두면 빠르게 처리한다.
    '''
    answer = 0
    sorted_costs = sorted(costs, key=lambda x: x[2])

    visited = set([0])
    while len(visited) < n:
        for i in range(len(sorted_costs)):
            v1, v2, cost = sorted_costs[i]
            if v1 in visited and v2 in visited:
                sorted_costs.pop(i)
                break
            elif v1 in visited or v2 in visited:
                visited |= set([v1, v2])
                answer += cost
                sorted_costs.pop(i)
                break

    return answer

# ---------------------------------------------
def find(parents, i):
    if parents[i] == i:
        return i
    else:
        return find(parents, parents[i])


def union(parents, node1, node2):
    parents[find(parents, node2)] = find(parents, node1)


def kruskal_solution(n, costs):
    '''
    https://gmlwjd9405.github.io/2018/08/29/algorithm-kruskal-mst.html
    https://yabmoons.tistory.com/186

    [ Kruskal Algorithm ]
    네트워크의 모든 정점을 최소 비용으로 연결하는 최적 해답 알고리즘
    MST(최소 비용 신장 트리)에 근거하여
    1. 최소 비용의 간선으로 구성
    2. 사이클을 포함하지 않음
    -------------------------------
    1. 그래프의 간선을 오름차순으로 정렬한다.
    2. 사이클을 만들지 않는 간선 중 최소 비용 간선을 선택한다.
       (union - find 방법 사용)

    * 주의 *
    parents 내 원소는 항상 tree의 root 원소를 의미하지 않지만
    find 함수를 거치면 root가 나온다.
    '''
    answer = 0
    parents = list(range(n))
    sorted_costs = sorted(costs, key=lambda x: x[2], reverse=True)

    # 모두 연결될 때까지(모두 뿌리가 같을 때) 실행
    while len(set(map(lambda x: find(parents, x), range(n)))) > 1:
        i1, i2, cost = sorted_costs.pop()
        if find(parents, i1) != find(parents, i2):
            union(parents, i1, i2)
            answer += cost

    return answer

File: programmers/algorithm_data_structure/greedy/섬_연결하기.java
package algorithm_data_structure.greedy;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class 섬_연결하기 {

    /* 
     * 크루스칼 알고리즘(+ find-union) 사용 풀이
     * find-union 방법 : https://gmlwjd9405.github.io/2018/08/31/algorithm-union-find.html
     * 
     * 최소 비용의 필요 조건은 최소 연결이므로
     * 연결 노드 수가 n-1개가 되면 수행 종료 
     */
    public int mySolution1(int n, int[][] costs) {
        int answer = 0;
        int connectCount = 0;
        int[] root = new int[n];
        int[] level = new int[n];

        List<int[]> sortedCosts = new ArrayList<>();

        for(int i = 0; i < n; i++) {
            root[i] = i;
        }
        for(int[] cost : costs) {
            sortedCosts.add(cost);
        }
        sortedCosts.sort((o1, o2) -> o1[2] - o2[2]);

        for (int[] cost : sortedCosts) {
            if (find(cost[0], root) != find(cost[1], root)) {
                union(cost[0], cost[1], root, level);
                answer += cost[2];
                connectCount += 1;
            }

            if (connectCount == n-1) {
                break;
            }
        }
        
        return answer;
    }
    
    public int find(int node, int[] root) {
        if (node != root[node]) {
            root[node] = find(root[node], root);
            return root[node];
        } else {
            return node;
        }
    }

    public void union(int node1, int node2, int[] root, int[] level) {
        int root1 = find(node1, root);
        int root2 = find(node2, root);

        if (level[root1] < level[root2]) {
            root[root1] = root2;
        } else if (level[root1] > level[root2]) {
            root[root2] = root1;
        } else {
            root[root2] = root1;
            level[root1] += 1;
        }
    }

    // 프림 알고리즘: 위 크루스칼 알고리즘과 비슷한 실행 시간
    public int mySolution2(int n, int[][] costs) {
        int answer = 0;
        Set<Integer> connected = new HashSet<>();
        connected.add(0);

        while (connected.size() < n) {
            int[] selected = null;

            for (int[] cost : costs) {
                if (connected.contains(cost[0]) && !connected.contains(cost[1])
                   || connected.contains(cost[1]) && !connected.contains(cost[0])) {
                    if (selected == null || selected[2] > cost[2]) {
                        selected = cost;
                    }
                }
            }

            connected.add(selected[0]);
            connected.add(selected[1]);
            answer += selected[2];
        }

        return answer;
    }

    // 프림 알고리즘 사용 2
    // costs 자료형을 매번 줄여나가지만
    // 줄이는 비용이 오히려 커서 앞의 코드보다 시간이 더 나감
    public int mySolution3(int n, int[][] costs) { 
        int answer = 0;
        List<int[]> copiedCosts = new ArrayList<>();
        for (int i = 0; i < costs.length; i++) {
            copiedCosts.add(costs[i]);
        }
        Set<Integer> connected = new HashSet<>();
        connected.add(0);

        while (connected.size() < n) {
            int[] selected = null;
            int selectedIndex = -1;

            for (int i = 0; i < copiedCosts.size(); i++) {
                int[] cost = copiedCosts.get(i);
                if (connected.contains(cost[0]) && !connected.contains(cost[1])
                        || connected.contains(cost[1]) && !connected.contains(cost[0])) {
                    if (selected == null || selected[2] > cost[2]) {
                        selected = cost;
                        selectedIndex = i;
                    }
                }
            }

            copiedCosts.remove(selectedIndex);
            connected.add(selected[0]);
            connected.add(selected[1]);
            answer += selected[2];
        }

        return answer;
    }
}